### PR TITLE
TRCL-3672 Static Typing: TradeInputValidator (and its child validators)

### PR DESCRIPTION
**Pull request by ruixhuang**
* Migrating TradeInputValidator (and its child validators).  Trade input validation should work now except for FieldInputValidator
* FieldInputValidator needs extra work since it relies on untyped lookup maps
* Will add unit tests and update integration tests in follow-up PRs
* Fixing bug with the validation error output sorting logic
* Tested on Android and guard by feature flag AppConfigsV2.staticTyping

### CODE REVIEW

File: src/commonMain/kotlin/exchange.dydx.abacus/output/input/TradeInput.kt [hunk 1]
```diff
@@ -735,7 +753,7 @@ enum class OrderType(val rawValue: String) {
 
     companion object {
         operator fun invoke(rawValue: String?) =
-            OrderType.values().firstOrNull { it.rawValue == rawValue }
+            entries.firstOrNull { it.rawValue == rawValue }
```
Comment (jaredvu) [inline src/commonMain/kotlin/exchange.dydx.abacus/output/input/TradeInput.kt:756]: entries === values()?

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt [hunk 2]
```diff
@@ -636,27 +636,29 @@ open class TradingStateMachine(
...
+            }
 
             if (subaccountNumber != null) {
                 if (staticTyping) {
                     when (internalState.input.currentType) {
                         InputType.TRADE -> {
                             calculateTrade(subaccountNumber)
                         }
-                        InputType.TRADE -> {
+                        InputType.TRANSFER -> {
```
Comment (jaredvu) [inline src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt:661]: 😄

File: src/commonMain/kotlin/exchange.dydx.abacus/validator/trade/TradeBracketOrdersValidator.kt [hunk 3]
```diff
@@ -366,6 +707,39 @@ internal class TradeBracketOrdersValidator(
...
     private fun validateStopLossReduceOnly(
+        trade: InternalTradeInputState,
+        position: InternalPerpetualPosition,
+    ): ValidationError? {
+        val reduceOnly = trade.brackets?.stopLoss?.reduceOnly ?: false
+        val sizePostOrder = position.calculated[CalculationPeriod.post]?.size ?: Numeric.double.ZERO
+        return if (reduceOnly) {
+            when (trade.side) {
+                OrderSide.Sell -> {
+                    if (sizePostOrder > Numeric.double.ZERO) {
```
Comment (jaredvu) [inline src/commonMain/kotlin/exchange.dydx.abacus/validator/trade/TradeBracketOrdersValidator.kt:718]: For validating `reduceOnly` shouldn't we also compare `sizePostOrder` with the `currentPositionSize`?

File: src/commonMain/kotlin/exchange.dydx.abacus/validator/trade/TradeBracketOrdersValidator.kt [hunk 4]
```diff
@@ -366,6 +707,39 @@ internal class TradeBracketOrdersValidator(
...
+                        reduceOnlyError(
+                            field = listOf("brackets.stopLoss.triggerPrice", "brackets.stopLoss.reduceOnly"),
+                        )
+                    } else {
+                        null
+                    }
+                }
+                OrderSide.Buy -> {
+                    if (sizePostOrder < Numeric.double.ZERO) {
```
Comment (jaredvu) [inline src/commonMain/kotlin/exchange.dydx.abacus/validator/trade/TradeBracketOrdersValidator.kt:727]: For validating reduceOnly shouldn't we also compare sizePostOrder with the currentPositionSize?

Is there a separate conditional that handles that and this is only for the case in which they increase a position that is crossing?

**Review by jaredvu — APPROVED** [review]

File: src/commonMain/kotlin/exchange.dydx.abacus/output/input/TradeInput.kt (hunk 1 above)
Comment (ruixhuang) [reply src/commonMain/kotlin/exchange.dydx.abacus/output/input/TradeInput.kt:756]: Yeah.. haha

File: src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt (hunk 2 above)
Comment (ruixhuang) [reply src/commonMain/kotlin/exchange.dydx.abacus/state/model/TradingStateMachine.kt:661]: lol... my bad

File: src/commonMain/kotlin/exchange.dydx.abacus/validator/InputValidator.kt [hunk 5]
```diff
@@ -15,8 +17,8 @@ internal class InputValidator(
     val parser: ParserProtocol,
 ) {
     private val errorTypeLookup = mapOf<String, Int>(
-        "ERROR" to 0,
-        "REQUIRED" to 1,
+        "REQUIRED" to 0,
+        "ERROR" to 1,
```
[the remark this answers is not in the document]
Comment (ruixhuang) [reply src/commonMain/kotlin/exchange.dydx.abacus/validator/InputValidator.kt:21]: Yes... This wasn't even called correctly so those errors are not shown.

File: src/commonMain/kotlin/exchange.dydx.abacus/validator/trade/TradeBracketOrdersValidator.kt (hunk 4 above)
Comment (ruixhuang) [reply src/commonMain/kotlin/exchange.dydx.abacus/validator/trade/TradeBracketOrdersValidator.kt:727]: Yes, it's in TradePositionStateValidator

**Review by jaredvu — APPROVED** [review]

**Pull request merged**